Rationale/spec improvements

mgorny opened the pull request:
Per the feedback, try to make a better split of rationale/specification.

Notably, at this point Rationale tries to introduce all the important concepts, so it's mostly standalone. After reading it, one should have an idea of all the important concepts, and need Specification only to learn the details of how to implement them.

I've also noticed we're not really covering the use of providers in sufficient detail in the specification section, so expanded on the appropriate MUSTs and SHOULDs.

Potential TODO: we may also want to mention environment markers and ABI namespace in Rationale.

mgorny committed:
Move glossary and overview into Rationale

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Remove speccy parts from "Overview" rationale

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Remove "summary of changes" for now

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Highlight the added filename part

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Avoid forward-linking from rationale to spec

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Expand variant properties in rationale

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Better explain providers in "Rationale" section

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Readd info on using providers to the spec

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Explain metadata a bit in rationale

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Clarify provider part of spec

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Explain ABI variant provider in the rationale

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Rewording of confusing bits

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Remove duplicate "publishes builds" verb

Signed-off-by: Michał Górny <mgorny@quansight.com>

rgommers commented on peps/pep-0817.rst:
```diff
@@ -716,36 +716,80 @@ build against.
+wheel maintainer or queried at wheel build time from an AoT plugin. Both
+kinds of plugins are usually implemented as Python packages which
+implement the provider plugin API, but they may also be vendored or
+reimplemented by installers to improve security, as outlined in TODO.
```
There's a TODO left here

rgommers commented on peps/pep-0817.rst:
```diff
@@ -1091,6 +1228,7 @@ otherwise processing it. To uphold this assumption, the proposal
 explicitly requires that untrusted provider plugin packages are never installed
 without explicit user consent.
 
+// TODO: update this
```
Another TODO

rgommers commented on peps/pep-0817.rst:
```diff
@@ -716,36 +716,80 @@ build against.
 Rationale
 =========
 
-Summary of changes
```
Was this duplicate, or incomplete? It seems pretty useful to have a concise summary of changes to guide the reader.

mgorny replied:
It was duplicate to the TOC, and it was pretty arbitrary to begin with. Like, it didn't really focus on what the spec is about, but rather on what you could easily bullet point.

mgorny replied:
Oh, and the "Overview" section now serves the same purpose. It's a little more verbose, but I think it explains the links better than a list.

rgommers replied:
Okay, seems reasonable. I'll leave the comment open, since others may have the same question.

rgommers commented on peps/pep-0817.rst:
```diff
@@ -867,39 +880,58 @@ need to be compatible. This provides some flexibility in designating
+A null variant is a variant wheel with no properties, but distinct
+from non-variant wheels in having the ``null`` variant label and variant
+metadata. During the transition period, it provides the possibility of
+providing a distinct fallbacks for systems that do not support any of
```
```suggestion
providing a distinct fallback for systems that do not support any of
```

rgommers commented on peps/pep-0817.rst:
```diff
@@ -868,56 +944,60 @@ distinguishable from regular variants.
-irrespective of the user system. All AoT providers can be technically
-used as install-time providers, but not the other way around.
+The variant wheel metadata specifies what providers are used for its
+properties. Providers serve twofold purpose:
```
```suggestion
properties. Providers serve a twofold purpose:
```

rgommers commented on peps/pep-0817.rst:
```diff
@@ -868,56 +944,60 @@ distinguishable from regular variants.
+providers and Ahead-of-Time providers.
+
+Install-time providers are implemented either as Python packages that
+need to be installed and ran to query them, or vendored or reimplemented
```
```suggestion
need to be installed and run to query them, or vendored or reimplemented
```

rgommers commented on peps/pep-0817.rst:
```diff
@@ -868,56 +944,60 @@ distinguishable from regular variants.
+running code external to the installer, they do not pose the problems
+faced by install-time providers, and can be used more liberally.
+
+AoT providers are permitted to feature plugin packages. If that is the
```
This first sentence seems confusing. It raises the question of what the alternative is, since they do need to be implemented somewhere and presumably that's in a plugin package.

mgorny replied:
"Provider" can be either a plugin package or inline metadata in `pyproject.toml`. That's why we've switched from saying "plugins".

mgorny replied:
Actually, the previous paragraph states that they are "implemented as static metadata embedded in            the wheel".

rgommers commented on peps/pep-0817.rst:
```diff
@@ -868,56 +944,60 @@ distinguishable from regular variants.
+file.
+
+When implemented as Python packages, both kinds of provider plugins
+roughly expose the same API. However, an AoT provider must always
```
```suggestion
expose roughly the same API. However, an AoT provider must always
```

rgommers commented on peps/pep-0817.rst:
```diff
@@ -868,56 +944,60 @@ distinguishable from regular variants.
+roughly expose the same API. However, an AoT provider must always
+consider all valid variant properties supported, and it must always
+return the same ordered list of supported properties irrespective of the
+user system. All AoT providers can be technically used as install-time
```
grammar:
```suggestion
user system. All AoT providers can technically be used as install-time
```

rgommers replied:
More importantly: is this useful? It seems more confusing than helpful. If it's an AoT provider then it doesn't do anything at install time by definition, so why would anyone do this?

mgorny replied:
I suppose the answer is "it depends".

It's pretty much a byproduct of having the same API for both kinds of plugins. And I think it makes sense to have a consistent API, if only to make things simpler. The only real difference is for install-time provider, `get_supported_configs()` is called at install time, whereas for AoT provider it's called at build time (hence Ahead-of-Time).

Technically, we could make the API more compact, say, by reusing `get_all_configs()` for that purpose and removing `get_supported_configs()` for AoT providers (after all, they return the same thing, modulo ordering). However, that will make the spec and the implementation harder. You'd have things like "for install-time providers, `get_supported_configs()` is used and `get_all_configs()` is unordered; for AoT providers, `get_all_configs()` is used and it needs to be ordered". So I don't think that's a good thing to do.

Now, why would anyone use an AoT provider as install-time provider: most of the time, that would probably be a mistake (not realizing they could AoT it). Perhaps we should say that builders SHOULD emit a warning if that happens.

Two kinda-valid use cases come to my mind:
1. Testing while developing a plugin — it's easier to work on it if it's queried at install time, rather than having to rebuild the test wheel for every change.
2. A past install-time plugin becoming AoT plugin — i.e. basically people realize they could make it AoT, but there are packages using it as install-time plugin already.

rgommers replied:
My point wasn't to suggest design changes, but more to point out that it doesn't seem useful to discuss this point here - the text is confusing, so why point it out at all here?

> Perhaps we should say that builders SHOULD emit a warning if that happens.

Either that or doing nothing both seem fine to me.

mgorny replied:
Ah. I was mostly trying here to strengthen the point that AoT providers have additional restrictions on them.

rgommers commented on peps/pep-0817.rst:
```diff
@@ -941,17 +1021,74 @@ the external plugin when the package in question is incompatible with
+Metadata in project tree and wheels
+-----------------------------------
+
+Variants introduce a few new portions of metadata that is stored in the
```
```suggestion
Variants introduce a few new portions of metadata that are stored in the
```

rgommers commented on peps/pep-0817.rst:
```diff
@@ -941,17 +1021,74 @@ the external plugin when the package in question is incompatible with
+-----------------------------------
+
+Variants introduce a few new portions of metadata that is stored in the
+project tree and in wheels. In the project tree, it is stored in the
```
"project tree" isn't a packaging concept, replace with "source tree"?

rgommers commented on peps/pep-0817.rst:
```diff
@@ -941,17 +1021,74 @@ the external plugin when the package in question is incompatible with
+into the inconvenient :doc:`Core Metadata
+<packaging:specifications/core-metadata>` format.
+
+The project tree-level metadata includes:
```
```suggestion
The metadata in ``pyproject.toml`` includes:
```

rgommers commented on peps/pep-0817.rst:
```diff
@@ -941,17 +1021,74 @@ the external plugin when the package in question is incompatible with
+that is used by clients to efficiently obtain the variant metadata
+without having to download it from individual wheels separately, or
+implement explicit variant metadata support on the package index
+software.
```
```suggestion
implement explicit variant metadata support in an API provided by the
package index server.
```

rgommers commented on peps/pep-0817.rst:
```diff
@@ -941,17 +1021,74 @@ the external plugin when the package in question is incompatible with
+ABI dependency variant provider
+-------------------------------
+
+Some packages provide extension modules exposing Application Binary
```
```suggestion
Some packages provide extension modules exposing an Application Binary
```

rgommers commented on peps/pep-0817.rst:
```diff
@@ -941,17 +1021,74 @@ the external plugin when the package in question is incompatible with
+narrow and users face problems if they need to install two packages that
+may happen to pin to different versions of the same dependency.
+Providing variants built against different dependency versions can
+increase the chances of resolver being able to find a dependency version
```
```suggestion
increase the chance of a resolver being able to find a dependency version
```

rgommers commented on peps/pep-0817.rst:
```diff
@@ -941,17 +1021,74 @@ the external plugin when the package in question is incompatible with
+
+Given the complexity of the problem, this extension is made entirely
+optional. This implies that any packages using it need to provide
+regular wheels as well.
```
I suggest "regular" -> "non-variant" (or "regular (non-variant)"). I assume in a few years variant wheels _are_ regular wheels.

rgommers commented on peps/pep-0817.rst:
```diff
@@ -1325,6 +1356,65 @@ Examples:
+Providers
+---------
+
+When installing variant wheels, installers SHOULD verify whether a given
```
This SHOULD seems to conflict with the MUST in the next paragraph. That's because this paragraph is about `install some-filename.whl` while the next paragraph is about `install pkgname` (right?). Would be nice to clarify that.

mgorny replied:
Yes, that was my thinking.

mgorny replied:
Though perhaps that's too clever. I'll change it to just SHOULD.

rgommers commented on peps/pep-0817.rst:
```diff
@@ -1325,6 +1356,65 @@ Examples:
+wheels are going to be installed. It SHOULD NOT use any providers whose
+markers do not match, and instead assume that its properties are
+incompatible. However, providers themselves MUST return no compatible
+properties, if queried on such a system.
```
This seems wrong. It's very well possible for a provider to work on some platform but a package using the provider to have excluded the installation of that provider on that same platform.

mgorny replied:
Hmm, that's a valid concern. My thinking was that environment markers were a "performance" thing, i.e. not having to query aarch64 provider on other platforms, but I suppose that's a valid use case.

rgommers commented on peps/pep-0817.rst:
```diff
@@ -1325,6 +1356,65 @@ Examples:
+All the tools that need to query variant providers and are run in a
+security-sensitive context, MUST NOT install or run code from any
+untrusted package for variant resolution without explicit user opt-in.
+Provider packages SHOULD take measures to guard against supply chain
```
Install-time provider packages? I don't quite see a reason to say that AoT providers SHOULD do this.

mgorny replied:
Makes sense.

rgommers commented on peps/pep-0817.rst:
```diff
@@ -1325,6 +1356,65 @@ Examples:
+untrusted package for variant resolution without explicit user opt-in.
+Provider packages SHOULD take measures to guard against supply chain
+attacks, for example by vendoring all dependencies. Pinning dependencies
+is discouraged to comply with :pep:`517` build process, as provider
```
From here onwards, this paragraph doesn't seem content for Specification. Maybe just omit it?

mgorny replied:
Indeed. Though I have mixed feelings about omitting it — the last thing we need are people pinning all dependencies and then getting reports that their plugins cannot be used together. Splitting things halfway between here and rationale also seems suboptimal.

mgorny replied:
Moved it to rationale, into a dedicated paragraph on how plugins are installed.

rgommers commented on peps/pep-0817.rst:
```diff
@@ -1325,6 +1356,65 @@ Examples:
+
+It is RECOMMENDED that said tools vendor, reimplement or lock the most
+commonly used plugins to specific wheels after verifying
+their trustworthiness, to enable the ability to securely install variant
```
"after verifying their trustworthiness" doesn't look like it belongs in a Specification section. Can just be omitted I think.

rgommers commented on peps/pep-0817.rst:
```diff
@@ -1325,6 +1356,65 @@ Examples:
+their trustworthiness, to enable the ability to securely install variant
+wheels out-of-the-box. To reduce the maintenance costs, repositories of
+such vetted plugins could be maintained collaboratively and shared
+between different tools.
```
All this belongs in Rationale I think, except for the first half of the first sentence of this paragraph.

mgorny replied:
Or maybe in "security considerations"?

mgorny replied:
Actually, it's already mentioned there, so I'll just remove it.

rgommers commented on peps/pep-0817.rst:
```diff
@@ -2215,6 +2312,12 @@ wheels or not, :pep:`517` and :pep:`660` hooks MUST build non-variant
+determine whether variant properties requested by the user are valid.
+They MUST refuse to build variant wheels with invalid metadata, and
+SHOULD refuse to build wheels with properties that are not considered
+valid according to their providers.
```
I don't quite understand the SHOULD/MUST/SHOULD here. How can the MUST be guaranteed if the "should" isn't followed? I.e., a build backend can't know if variant metadata is valid if it doesn't query the relevant provider.

mgorny replied:
It's the distinction between "invalid metadata" vs. "invalid properties". "Invalid metadata" are things like incorrect TOML keys.

mgorny replied:
I reordered it and clarified that we're talking of "nonconformant" metadata vs "unknown" properties.

rgommers reviewed (commented):
Thanks @mgorny. A lot more feedback on this one

mgorny committed:
Apply suggestions from code review

Co-authored-by: Ralf Gommers <ralf.gommers@gmail.com>

mgorny committed:
project tree → source tree

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
regular wheel → non-variant wheel

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Loosen and clarify verification at install time

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Make respecting enable-if mandatory

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Limit recommendations to install-time providers

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Remove "after verifying their trustworthiness"

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Try to clarify build-time validation

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Update links to where security is covered now

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Remove duplicate/non-spec content in Providers spec

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Move environment and pinning bits to rationale

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Mention metadata bits in overview

Signed-off-by: Michał Górny <mgorny@quansight.com>

mgorny committed:
Remove some duplication in "Provider information"

Environment marker usage and optionality is described in "Providers"
section. Let's limit "Provider information" only to technical field
details themselves.

Signed-off-by: Michał Górny <mgorny@quansight.com>

rgommers reviewed (approved):
LGTM now, thanks @mgorny
